File: leetcode/Medium/49.GroupAnagrams.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

// 26ms

func groupAnagrams(strs []string) [][]string {
	mp := make(map[string][]string)
	for _, s := range strs {
		tmp := strings.Split(s, "")
		sort.Strings(tmp)
		sorted_s := strings.Join(tmp, "")
		mp[sorted_s] = append(mp[sorted_s], s)
	}

	var res [][]string
	for _, vec := range mp {
		res = append(res, vec)
	}
	return res
}

func main() {
	vec := []string{"eat", "tea", "tan", "ate", "nat", "bat"}
	fmt.Println(groupAnagrams(vec))
}

File: leetcode/Medium/49.GroupAnagrams.cpp
#include <bits/stdc++.h>
using namespace std;
#define _GLIBCXX_DEBUG

// 44ms

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        map<string, vector<string>> mp;
        for (string &s : strs) {
            string sorted_s = s;
            sort(sorted_s.begin(), sorted_s.end());
            mp[sorted_s].push_back(s);
        }

        vector<vector<string>> res;
        for (auto &[_, vec] : mp) {
            res.push_back(vec);
        }
        return res;
    }
};
